rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
+            subscriptions,
+            subscriptions_buffers: RefCell::new(heapless::Vec::new()),
+            buffers,
+        }
+    }
+
+    /// Answer a responding exchange using the `DataModelHandler` instance wrapped by this exchange handler.
+    pub async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let mut timeout_instant = None;
+
+        loop {
+            let mut repeat = false;
+
+            if exchange.rx().is_err() {
+                exchange.recv_fetch().await?;
+            }
+
+            let meta = exchange.rx()?.meta();
+            if meta.proto_id != PROTO_ID_INTERACTION_MODEL {
+                Err(ErrorCode::InvalidProto)?;
+            }
+
+            match meta.opcode::<OpCode>()? {
+                OpCode::ReadRequest => self.read(exchange).await?,
+                OpCode::WriteRequest => {
+                    repeat = self.write(exchange, timeout_instant.take()).await?;
+                }
+                OpCode::InvokeRequest => self.invoke(exchange, timeout_instant.take()).await?,
+                OpCode::SubscribeRequest => self.subscribe(exchange).await?,
+                OpCode::TimedRequest => {
+                    timeout_instant = Some(self.timed(exchange).await?);
+                    repeat = true;
+                }
+                opcode => {
+                    error!("Invalid opcode: {:?}", opcode);
+                    Err(ErrorCode::InvalidOpcode)?
+                }
+            }
 
-pub struct DataModel<T>(T);
+            if !repeat {
+                break;
+            }
+        }
 
-impl<T> DataModel<T> {
-    pub fn new(handler: T) -> Self {
-        Self(handler)
+        exchange.acknowledge().await?;
+        exchange.matter().notify_changed();
+
+        Ok(())
     }
 
-    pub async fn handle<'r, 'p>(
+    async fn read(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let Some(mut tx) = self.tx_buffer(exchange).await? else {
+            return Ok(());
+        };
+
+        let mut wb = WriteBuf::new(&mut tx);
+
+        let metadata = self.handler.lock().await;
+
+        let req = ReadReq::from_tlv(&get_root_node_struct(exchange.rx()?.payload())?)?;
+        debug!("IM: Read request: {:?}", req);
+
+        let req = ReportDataReq::Read(&req);
+
+        let accessor = exchange.accessor()?;
+
+        let awaits = metadata.node().read(&req, None, &accessor).any(|item| {
+            item.map(|attr| self.handler.read_awaits(&attr))
+                .unwrap_or(false)
+        });
+
+        if !awaits {
+            let node = metadata.node();
+            let mut attrs = node.read(&req, None, &accessor).peekable();
+
+            if !req
+                .respond(&self.handler, None, &mut attrs, &mut wb, true)
+                .await?
+            {
+                drop(attrs);
+
+                exchange.send(OpCode::ReportData, wb.as_slice()).await?;
+
+                // Self::recv_status(exchange).await?;
```

Comment:
consider deleting commented-out code.

Reply:
For now, I've put a comment as to why this line is commented out and what is the issue at hand in the first place.